### PR TITLE
Travis cache

### PR DESCRIPTION
**Pull request by frabbit**
this PR enabled caching for some folders during the build process, the following folders are cached and reused between builds
- .opam - opam files
- neko - the neko folder including compiled files
- apt-cache - downloaded apt-get files (linux)
- hxcache - hxcpp cache
- _hxbuild - the compiler _build folder
- lua_env - lua files
- .luarocks - .luarocks folder (@jdonaldson please check)
- Library/Caches/Homebrew - downloaded homebrew files (osx)

Although it is not recommended to cache apt-get or homebrew files, it seems to make a difference in my tests.

Sorry for the stupid commits, but it's hard to test travis files ;)

### CODE REVIEW

**Comment by Simn** [comment]
@andyli's birthday present. :D

**Comment by frabbit** [comment]
at least the commits show that it was more complicated and annoying as someone would assume ;)

**Comment by frabbit** [comment]
@justin, i wonder if it's somehow possible to avoid or cache the installation of the lua jits, i tried to cache $HOME/lua_env but it's still recompiled by luarocks (luarocks install ...) in some temp directory every time we run a build, maybe you can find a nice way to cache the compiled files or the binary itself, i think we can at least save ~3 minutes for each build

File: .travis.yml [hunk 1]
```diff
@@ -42,8 +62,15 @@ install_linux: &install_linux
...
   # Install neko
-  - git clone https://github.com/HaxeFoundation/neko.git ../neko
-  - pushd ../neko
+  - if [ ! -d "$HOME/neko" ]; then
+      mkdir $HOME/neko;
+      pushd $HOME/neko;
+      git clone https://github.com/HaxeFoundation/neko.git .;
+    else
+      pushd $HOME/neko;
+      git pull origin master || git clone https://github.com/HaxeFoundation/neko.git .;
```
Comment (andyli) [inline .travis.yml]: I think a `git fetch --all; git reset --hard origin/master` would be more reliable when there is merge conflict.

**Comment by frabbit** [comment]
@jdonaldson i think i got it, can you check this commit https://github.com/HaxeFoundation/haxe/pull/6366/commits/6d1b29f212e90059c5f608638bd2abb689c13dd7

File: .travis.yml (hunk 1 above)
Comment (frabbit) [inline .travis.yml]: sounds good

**Comment by frabbit** [comment]
@andyli i still have a problem with the osx timeout (50m) when we don't have a cache, but the cache is job specific, so i can't run a different job to build the cache or something.

**Comment by frabbit** [comment]
@andyli even target splitting is not an option for osx, because cpp alone takes longer than 50m

**Comment by andyli** [comment]
@hughsando Is there a hxcpp flag or something that can specify the number of threads used for compiling?

**Comment by ibilon** [comment]
> @hughsando Is there a hxcpp flag or something that can specify the number of threads used for compiling?

There's `HXCPP_COMPILE_THREADS`, but by default it's the number of cores.

Is hxcpp cache used, and cached?

**Comment by frabbit** [comment]
ok, got an idea, i just fail on osx before running the script when there is no cache, so that at least the cache for neko/brew/opam etc. is created, in that case you would have to restart the build which is a fair limitation i guess.

**Comment by andyli** [comment]
> There's HXCPP_COMPILE_THREADS, but by default it's the number of cores.

We may want to check the actual number of thread being used since the cpu core number check could be unreliable in visualized environment, so it could be using more threads than optimal.

There should be [2 cores](https://docs.travis-ci.com/user/ci-environment/#Virtualization-environments) available, so 4 threads should be good.

**Comment by frabbit** [comment]
@andyli are you suggesting setting HXCPP_COMPILE_THREADS=4 in every case?

**Comment by andyli** [comment]
Yes. But it could be already using 4 threads.
Should turn on verbose mode to check: https://github.com/HaxeFoundation/hxcpp/blob/9b8f24f1e7144fb4fe6d21e081d5f8440aba055b/tools/hxcpp/BuildTool.hx#L313

**Comment by frabbit** [comment]
let's wait and see...
@andyli any tips how to debug travis scripts locally or faster? ;)

**Comment by frabbit** [comment]
ok when a build failed the cache is not created/updated. I'm looking for a way to create the cache without saying that the build was successfull (exit 0).

**Comment by andyli** [comment]
> any tips how to debug travis scripts locally or faster? ;)

Only possible when using `sudo: false`, which is not our case...
https://docs.travis-ci.com/user/common-build-problems/#Troubleshooting-Locally-in-a-Docker-Image


AppVeyor allows remote desktop into the build: https://www.appveyor.com/docs/how-to/rdp-to-build-worker/
Though there is no official support of such thing in Travis, I guess it is possible...

**Comment by frabbit** [comment]
Ok, finally this should now work as expected. On OSX we only create the cache (at least a part of it) when it's not available and fail with a no-cache message. Only when the cache is available the tests are run. Now we should have faster builds and no more timeouts on OSX. But OSX jobs will still take forever until they start because of the travis infrastructure.

**Comment by frabbit** [comment]
image

**Comment by frabbit** [comment]
Ready to merge

**Pull request merged**